### ISSUE TITLE
OTA Firmware Updates Can Stall

### ISSUE THREAD

**Issue by eely22** [bluzDK/bluzDK-firmware#40]
OTA Firmware Updates can occasionally stall, that is they never complete and seemingly get stuck and eventually hit the max timeout and reboot.

The issue is extremely intermittent and random, and it is very hard to reproduce and debug. This likely means a race condition or some other timing issue.

It usually doesn't happen on user app updates as they are small, it is much more noticeable during system updates which are much larger in size. The simple recovery is to do the update again, so the issue isn't blocking but can become very annoying very quickly.

**Comment by eely22** [comment]
I believe this may be solved. After much debugging and trial/error, I zoned in on the issue being in the bluz implementation of the socket class. I think there was a race condition that could happen when the interrupt to fill the socket buffer would run into the main thread moving through the receive function. I am not sure of the reason behind this, I can't logically trace a path to cause the problem, but through the process of elimination that seems to be the root. The really simple fix is to not move through the receive function. Also, some marked some variables that should have been marked volatile.

The issue is only in the transferring of the file, so I created a test that would wait for a device to come online and then transfer a dummy file. This would loop infinitely, creating a good stress test for the system. By commenting out the line that would reboot bluz in case of an OTA hang, I can reliably test this over long periods and across multiple gateway types.

Previously I could not get through a night with this test without hitting the issue. I am now running on many hours and things seem to still be working normally.

I am not closing this issue until the fix has been out in the wild and I get feedback that it is no longer seen.
